### PR TITLE
Recover from malformed operation implementation

### LINKED ISSUE #369: Unhelpful error message on missing braces in an interface implementation
When trying to write an instance of an interface:
```scala
interface Eff {
  def op(): Int
}

def e: Eff = new Eff {
  def op() = println("42"); 42
}
```

shows an error: "Operation Eff is not part of interface Eff."

Here's a screenshot:
Screenshot 2024-01-23 at 17 19 27

One way to work around this is to add braces into the method body:
```scala
interface Eff {
  def op(): Int
}

def e: Eff = new Eff {
  def op() = { println("42"); 42 }
}
```

### PR DESCRIPTION
**Pull request by jiribenes**
Resolves #369 

Before:
image

After:
image

### CODE REVIEW

File: effekt/shared/src/main/scala/effekt/RecursiveDescent.scala [hunk 1]
```diff
@@ -711,6 +724,24 @@ class RecursiveDescent(positions: Positions, tokens: Seq[Token], source: Source)
     nonterminal:
       (`def` ~> idRef()) ~ paramsOpt() ~ maybeReturnAnnotation() ~ (`=` ~> stmt()) match {
         case id ~ (tps, vps, bps) ~ ret ~ body =>
+         if (isSemi) {
+           semi()
+
+           val startPosition = position
+
+           // TODO(jiribenes): would be nice to abstract this and have a `consumeUntil` function
```
Comment (jiribenes) [inline effekt/shared/src/main/scala/effekt/RecursiveDescent.scala:732]:
```suggestion
```

File: effekt/shared/src/main/scala/effekt/RecursiveDescent.scala [hunk 2]
```diff
@@ -29,7 +32,17 @@ class RecursiveDescent(positions: Positions, tokens: Seq[Token], source: Source)
       //val after = System.currentTimeMillis()
       //println(s"${input.source.name}: ${after - before}ms")
 
-      res
+      // Report soft fails
+      softFails.foreach {
+        case SoftFail(msg, from, to) =>
+          val source = input.source
+          val fromPos = source.offsetToPosition(tokens(from).start)
+          val toPos = source.offsetToPosition(tokens(to).end) // NOTE(jiribenes): I think this is correct, we don't need the +1, right?
```
Comment (jiribenes) [inline effekt/shared/src/main/scala/effekt/RecursiveDescent.scala:40]:
```suggestion
          val toPos = source.offsetToPosition(tokens(to).end)
```

**Review by jiribenes — COMMENTED** [review]
The TODOs are okay 👀

**Review by dvdvgt — APPROVED** [review]
LGTM for this special error

File: effekt/shared/src/main/scala/effekt/RecursiveDescent.scala [hunk 3]
```diff
@@ -711,6 +724,23 @@ class RecursiveDescent(positions: Positions, tokens: Seq[Token], source: Source)
...
+           val startPosition = position
+
+           if (!peek(`}`) && !peek(`def`) && !peek(EOF)) {
+              // consume until the next `def` or `}` or EOF
+              while (!peek(`}`) && !peek(`def`) && !peek(EOF)) {
+                next()
+              }
+
+              val endPosition = position
+              val msg = "Unexpected tokens after operation definition. Expected either a new operation definition or the end of the implementation."
```
Comment (jiribenes) [inline effekt/shared/src/main/scala/effekt/RecursiveDescent.scala:739]: Ideally, we would have some hint with a quick fix like "did you mean to do `def op() = { ... }`?", but we don't have the infrastructure for this.

**Pull request merged**